src/tidy.zig

```diff
@@ -189,16 +202,99 @@ fn tidy_control_characters(file: SourceFile) ?u8 {
     return null;
 }
 
+const identifiers_per_file_max = 100_000;
+/// Detects unused constants and functions.
+///
+/// This is a one-side heuristics: there might be false negatives, but no false positives.
+///
+/// Current algorithm:
+/// - Two passes.
+/// - Pass 1: count how many times each identifier is mentioned in the file.
+/// - Pass 2: warn about any unique identifier which is a non-public declaration.
+///
+/// At the moment, this is implemented using only the lexer, without looking at the AST, as that
+/// seemed simpler.
+fn tidy_dead_declarations(
+    tree: *const std.zig.Ast,
+    used: *std.StringHashMapUnmanaged(u32),
+) ?[]const u8 {
+    assert(used.count() == 0);
+    defer used.clearRetainingCapacity();
+
+    var identifier_start: ?std.zig.Ast.ByteOffset = 0;
+    inline for (.{ .fill, .check }) |phase| {
+        next_token: for (
+            tree.tokens.items(.tag),
+            tree.tokens.items(.start),
+            0..,
+        ) |tag, start, index| {
+            const identifier_start_previous = identifier_start;
+            identifier_start = switch (tag) {
+                .identifier => start,
+                else => null,
+            };
+
+            const start_previous = identifier_start_previous orelse continue :next_token;
+            const token_text = std.mem.trim(
+                u8,
+                tree.source[start_previous..start],
+                &std.ascii.whitespace,
+            );
+
+            switch (phase) {
+                .fill => {
+                    const gop = used.getOrPutAssumeCapacity(token_text);
+                    if (!gop.found_existing) gop.value_ptr.* = 0;
+                    gop.value_ptr.* += 1;
+                    if (used.count() >= identifiers_per_file_max) @panic("file to large");
+                    continue :next_token;
+                },
+                .check => {
+                    const usages = used.get(token_text).?;
+                    assert(usages >= 1);
+                    if (usages > 1) continue :next_token;
+                },
+                else => comptime unreachable,
+            }
+
+            assert(phase == .check and used.get(token_text).? == 1);
```

Comment:
Why not split compound condition?:
```suggestion
            assert(phase == .check);
            assert(used.get(token_text).? == 1);
```

Reply:
Thanks, haven't quite internalized this rule yet!